### DISCUSSION TITLE
ZFS Syncoid : merge two sets of snapshots together  (feature request)

### DISCUSSION THREAD

**Discussion by MynaITLabs** [jimsalterjrs/sanoid#906] [Ideas]
Would be handy to help merge two sets of ZFS snapshots together into one clean separate filesystem with all snapshots.

For example,  
If there are 3 servers, each have the same ZFS version.
`MAIN`, `BACKUP1`, `BACKUP2`
* Syncoid does a daily and hourly snapshot `MAIN` server data files `/main-pool/data`  (in ZFS filesystem `main-pool/data`)
* Syncoid is daily transferring snapshots `main-pool` to `BACKUP1` zfs pool `backup1-pool/main-pool/data`.  (i.e. offsite backup)
* Snapshots older than 30 days are removed.  (Since they are actually on the `BACKUP1` server)
After 3 months, offsite server room fire damages the computer and SSD drives.  Bummer !
* Replacement and data recovery will take at least 42 days.
So I create a new server `BACKUP2`.
* Syncoid is now daily transferring snapshots `MAIN` 'main-pool' to `BACKUP2` zfs pool `backup2-pool/main-pool/data`. 
After 69 days, `BACKUP1` SSD drive data has been recovered, and back up and running
I now have two older snapshot sets that I want to merge together into one complete set.
* `BACKUP1` `backup1-pool/mail-pool/data`  (first 90 days of MAIN)
* `BACKUP2` `backup2-pool/mail-pool/data`  (day 91-160 of MAIN)

One method is make a list of all snapshots and sort by date, then create a new filesystem based on the oldest to newest.

Would be great if syncoid could take several parameters and do the 'hand-waving with smoke-and-mirrors' automagically.
Something like...
`$ syncoid --merge-source  BACKUP1:backup1-pool/mail-pool/data --merge-source  BACKUP2:backup2-pool/mail-pool/data   --merge-dest BACKUP1:backup1-pool/mail-pool/data` 

NOTE: 
*  if --merge-dest matches any of the sources, then would need a temp working area.  Then when complete process rename the DEST  to DEST_PREV and DEST_TEMP to DEST.

**Comment by MynaITLabs** [comment]
For a more complex examples: 
*  `BACKUP1` and `BACKUP2` actually had a mix of different days on them and not exactly sequential to each other ?
* There are several different sources of similar snapshots (`BACKUP1`, `BACKUP2`, `MAIN`, `BACKUP3`, `ZAPHOD`, etc)  merged into one dataset filesystem ?

**Comment by jimsalterjrs** [comment]
What you're asking for isn't really possible in ZFS, let alone sanoid/syncoid.

The closest thing to what you're looking for would be to make certain you do NOT delete the last common snapshot MAIN1 had with BACKUP1 before you send BACKUP1 off to Drivesavers or whoever. Then you build your BACKUP2. Then when BACKUP1 returns from the drive recovery outfit--which, honestly, it's probably not going to in the way that you think it is; they typically are only going to recover files from the main filesystem, not restore every snapshot--but assuming you *did*, in fact, get your actual original BACKUP1 system back with its pool intact, you could then simply replicate BACKUP2--which will still contain the last common snapshot between MAIN1 and BACKUP1--onto BACKUP1.

At that point, you just back up directly from MAIN1 to BACKUP1 again, and you tear down BACKUP2 for parts, repurpose it for something else, whatever.

Your second idea, of mixing and matching snapshots that are "not exactly sequential" and "on a mix of different days"--is literally impossible to do with ZFS replication, sorry.

**Comment by memnochproxy** [comment]
thanks for your input on this, Jim
